File: src/main/java/com/collicode/algodatastructure/leetcode/TwoSum.java
package com.collicode.algodatastructure.leetcode;

import java.util.HashMap;
import java.util.Map;

//SORTED ARRAY
class Solution {
    public int[] twoSum(int[] numbers, int target) {

        int left = 0;
        int right = numbers.length - 1;

        while (left < right) {
            int sum = numbers[left] + numbers[right];
            if (sum > target) {
                right--;
            } else if (sum < target) {
                left++;
            } else {
                return new int[]{left + 1, right + 1};
            }
        }

        return null;
    }
}

//UNSORTED  ARRAY
class Solution2 {
    public int[] twoSum(int[] numbers, int target) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < numbers.length; i++) {
            int complement = target - numbers[i];

            if (map.containsKey(complement)) {
                return new int[]{map.get(complement) + 1, i + 1};
            }
            map.put(numbers[i], i);
        }
        return null;
    }
}

File: src/main/java/com/collicode/algodatastructure/prep/week1/kotlin/TwoSum.kt
package com.collicode.algodatastructure.prep.week1.kotlin


class Solution {

    fun twoSum(numbers: IntArray, target: Int): IntArray? {
        var left = 0
        var right = numbers.size - 1

        while (left < right) {
            val sum = numbers[left] + numbers[right]

            when {
                sum > target -> right--
                sum < target -> left++
                else -> return intArrayOf(left + 1, right + 1)
            }
        }

        return null
    }
}



class Solution2 {

    fun twoSum(numbers: IntArray, target: Int): IntArray? {
        val map = mutableMapOf<Int, Int>()

        for (i in numbers.indices) {
            val complement = target - numbers[i]

            if (map.containsKey(complement)) {
                return intArrayOf(map[complement]!! + 1, i + 1)
            }

            map[numbers[i]] = i
        }

        return null
    }
}